File: qgis-plugin/grid_interoperability/adapters/s2.py
# -*- coding: utf-8 -*-
"""S2アダプター（s2sphere ラッパー）。

ID形式: S2セルトークン（16進、例 "60188bfc"）。レベル 0-30。
Web版 src/lib/adapters/s2.ts と同一の辺分割数・面積アルゴリズムを使う。
"""

import math
from typing import List, Optional

import s2sphere

from ..core.geometry import distance_m, polygon_area_m2
from ..core.models import GridCell

SYSTEM = "s2"
DEFAULT_LEVEL = 13
MAX_LEVEL = 30

# 1辺あたりの分割数（Web版 SUBDIV と同値にすること）
SUBDIV = 4


def _token_to_id(token: str) -> s2sphere.CellId:
    cid = s2sphere.CellId.from_token(token.strip())
    if not cid.is_valid():
        raise ValueError("無効なS2セルトークンです: %s" % token)
    return cid


def _interpolate(a, b, t):
    """球面上の2点間を弦補間して正規化（Web版と同一の近似）."""
    x = a[0] + (b[0] - a[0]) * t
    y = a[1] + (b[1] - a[1]) * t
    z = a[2] + (b[2] - a[2]) * t
    n = math.sqrt(x * x + y * y + z * z)
    lat = math.degrees(math.asin(max(-1.0, min(1.0, z / n))))
    lng = math.degrees(math.atan2(y, x))
    return (lng, lat)


def _unwrap_ring(ring, center_lon):
    out = []
    for lon, lat in ring:
        x = lon
        while x - center_lon > 180:
            x -= 360
        while x - center_lon < -180:
            x += 360
        out.append((x, lat))
    return out


def _cell_ring(cid: s2sphere.CellId):
    """セル境界リング（辺を SUBDIV 分割、閉じたリング）."""
    cell = s2sphere.Cell(cid)
    verts = [cell.get_vertex(k) for k in range(4)]
    ring = []
    for k in range(4):
        a, b = verts[k], verts[(k + 1) % 4]
        for i in range(SUBDIV):
            ring.append(_interpolate(a, b, i / SUBDIV))
    ring.append(ring[0])
    center = cid.to_lat_lng()
    return _unwrap_ring(ring, center.lng().degrees)


def _build_cell(cid: s2sphere.CellId) -> GridCell:
    level = cid.level()
    ll = cid.to_lat_lng()
    center = (ll.lng().degrees, ll.lat().degrees)
    ring = _cell_ring(cid)
    edge_sum = sum(
        distance_m(ring[k * SUBDIV], ring[((k + 1) % 4) * SUBDIV])
        for k in range(4)
    )
    return GridCell(
        system=SYSTEM,
        id=cid.to_token(),
        level=level,
        rings=[ring],
        center=center,
        area_m2=polygon_area_m2([ring]),
        edge_length_m=edge_sum / 4,
        parent_id=cid.parent(level - 1).to_token() if level > 0 else None,
        child_count=4 if level < MAX_LEVEL else 0,
    )


def point_to_cell(lon: float, lat: float, level, height_m: float = 0.0) -> GridCell:
    leaf = s2sphere.CellId.from_lat_lng(s2sphere.LatLng.from_degrees(lat, lon))
    return _build_cell(leaf.parent(int(level)))


def cell_to_geometry(token: str) -> GridCell:
    return _build_cell(_token_to_id(token))


def get_parent(token: str) -> Optional[str]:
    cid = _token_to_id(token)
    return cid.parent(cid.level() - 1).to_token() if cid.level() > 0 else None


def get_children(token: str) -> List[str]:
    cid = _token_to_id(token)
    if cid.level() >= MAX_LEVEL:
        return []
    return [c.to_token() for c in cid.children()]


def get_neighbors(token: str) -> List[str]:
    return [n.to_token() for n in _token_to_id(token).get_edge_neighbors()]


def _ring_bounds(ring):
    lons = [p[0] for p in ring]
    lats = [p[1] for p in ring]
    return (min(lons), min(lats), max(lons), max(lats))


def _boxes_intersect(a, b):
    lon_overlap = any(
        a[0] + s <= b[2] and a[2] + s >= b[0] for s in (-360, 0, 360)
    )
    return lon_overlap and a[1] <= b[3] and a[3] >= b[1]


def _cells_for_box(box, level, collected, max_cells):
    """中心セルから辺隣接BFSで範囲交差セルを列挙（Web版と同一ロジック）."""
    min_lon, min_lat, max_lon, max_lat = box
    seed = s2sphere.CellId.from_lat_lng(
        s2sphere.LatLng.from_degrees(
            (min_lat + max_lat) / 2, (min_lon + max_lon) / 2
        )
    ).parent(level)
    queue = [seed]
    visited = {seed.to_token()}
    while queue:
        cid = queue.pop(0)
        cell = _build_cell(cid)
        if not _boxes_intersect(_ring_bounds(cell.rings[0]), box):
            continue
        collected[cell.id] = cell
        if len(collected) > max_cells:
            raise ValueError("セル数が多すぎます。レベルを下げてください。")
        for n in cid.get_edge_neighbors():
            token = n.to_token()
            if token not in visited:
                visited.add(token)
                queue.append(n)


def cells_for_bounds(bounds, level) -> List[GridCell]:
    min_lon, min_lat, max_lon, max_lat = bounds
    collected = {}
    if min_lon <= max_lon:
        boxes = [(min_lon, min_lat, max_lon, max_lat)]
    else:
        # 国際日付変更線をまたぐ範囲は分割する
        boxes = [(min_lon, min_lat, 180.0, max_lat), (-180.0, min_lat, max_lon, max_lat)]
    for box in boxes:
        _cells_for_box(box, int(level), collected, 1000000)
    return list(collected.values())


def levels_for_ui():
    return [(lv, "level %d" % lv) for lv in range(MAX_LEVEL + 1)]

File: web/src/lib/adapters/s2.ts
/**
 * S2 アダプター（s2js ラッパー）。
 * ID形式: S2セルトークン（16進、例 "60188bfc"）。レベル 0-30。
 *
 * セル辺は大円のため、各辺を SUBDIV 分割して球面上を補間した頂点列で
 * ポリゴンを近似する。QGIS版 adapters/s2.py と同一の分割数・同一の
 * 面積アルゴリズムを使い、計算結果を一致させる（指示書 §16）。
 */

import { s2 } from "s2js";
import type { GridAdapter, GridCell, Position } from "../types";
import { distanceM, polygonAreaM2 } from "../geo";

const { cellid, Cell, LatLng } = s2;

const R2D = 180 / Math.PI;

/** 1辺あたりの分割数（QGIS版と同値にすること） */
export const SUBDIV = 4;

const MAX_LEVEL = 30;

function tokenToId(token: string): bigint {
  const id = cellid.fromToken(token.trim());
  if (!cellid.valid(id)) throw new Error(`無効なS2セルトークンです: ${token}`);
  return id;
}

/** 球面上の2点間を弦補間して正規化（両実装で同一の近似） */
function interpolate(
  a: { x: number; y: number; z: number },
  b: { x: number; y: number; z: number },
  t: number
): [number, number] {
  const x = a.x + (b.x - a.x) * t;
  const y = a.y + (b.y - a.y) * t;
  const z = a.z + (b.z - a.z) * t;
  const n = Math.sqrt(x * x + y * y + z * z);
  const lat = Math.asin(Math.max(-1, Math.min(1, z / n))) * R2D;
  const lng = Math.atan2(y, x) * R2D;
  return [lng, lat];
}

function unwrapRing(ring: Position[], centerLon: number): Position[] {
  return ring.map(([lon, lat]) => {
    let x = lon;
    while (x - centerLon > 180) x -= 360;
    while (x - centerLon < -180) x += 360;
    return [x, lat] as Position;
  });
}

/** セル境界リング（辺を SUBDIV 分割、閉じたリング） */
export function cellRing(id: bigint): Position[] {
  const cell = Cell.fromCellID(id);
  const verts = [0, 1, 2, 3].map((k) => {
    const v = cell.vertex(k) as unknown as {
      vector: { x: number; y: number; z: number };
    };
    return v.vector;
  });
  const ring: Position[] = [];
  for (let k = 0; k < 4; k++) {
    const a = verts[k];
    const b = verts[(k + 1) % 4];
    for (let i = 0; i < SUBDIV; i++) {
      ring.push(interpolate(a, b, i / SUBDIV));
    }
  }
  ring.push(ring[0]);
  const c = cellid.latLng(id) as unknown as { lat: number; lng: number };
  return unwrapRing(ring, c.lng * R2D);
}

function buildCell(id: bigint): GridCell {
  const level = cellid.level(id);
  const c = cellid.latLng(id) as unknown as { lat: number; lng: number };
  const center: [number, number] = [c.lng * R2D, c.lat * R2D];
  const ring = cellRing(id);
  // 4頂点間の測地距離の平均を辺長とする
  let edgeSum = 0;
  for (let k = 0; k < 4; k++) {
    edgeSum += distanceM(ring[k * SUBDIV], ring[((k + 1) % 4) * SUBDIV]);
  }
  return {
    system: "s2",
    id: cellid.toToken(id),
    level,
    geometry: { type: "Polygon", coordinates: [ring] },
    center,
    areaM2: polygonAreaM2([ring]),
    edgeLengthM: edgeSum / 4,
    parentId:
      level > 0 ? cellid.toToken(cellid.parent(id, level - 1)) : undefined,
    childCount: level < MAX_LEVEL ? 4 : 0,
  };
}

function ringBounds(ring: Position[]): [number, number, number, number] {
  let minLon = Infinity,
    minLat = Infinity,
    maxLon = -Infinity,
    maxLat = -Infinity;
  for (const [lon, lat] of ring) {
    if (lon < minLon) minLon = lon;
    if (lat < minLat) minLat = lat;
    if (lon > maxLon) maxLon = lon;
    if (lat > maxLat) maxLat = lat;
  }
  return [minLon, minLat, maxLon, maxLat];
}

function boxesIntersect(
  a: [number, number, number, number],
  b: [number, number, number, number]
): boolean {
  // 経度は±360ずらしても判定（アンラップ済みリング対策）
  const lonOverlap = [-360, 0, 360].some(
    (s) => a[0] + s <= b[2] && a[2] + s >= b[0]
  );
  return lonOverlap && a[1] <= b[3] && a[3] >= b[1];
}

/**
 * 範囲内セルの列挙: 中心セルから辺隣接をBFSで辿り、
 * 境界矩形が範囲と交差するセルを集める（RegionCoverer非依存の共通実装）。
 */
function cellsForBox(
  box: [number, number, number, number],
  level: number,
  collected: Map<string, GridCell>,
  maxCells: number
): void {
  const [minLon, minLat, maxLon, maxLat] = box;
  const seed = cellid.parent(
    cellid.fromLatLng(
      LatLng.fromDegrees((minLat + maxLat) / 2, (minLon + maxLon) / 2)
    ),
    level
  );
  const queue: bigint[] = [seed];
  const visited = new Set<string>([cellid.toToken(seed)]);
  while (queue.length) {
    const id = queue.shift()!;
    const cell = buildCell(id);
    const ring = cell.geometry.coordinates[0] as Position[];
    if (!boxesIntersect(ringBounds(ring), box)) continue;
    collected.set(cell.id, cell);
    if (collected.size > maxCells) {
      throw new Error("セル数が多すぎます。レベルを下げてください。");
    }
    for (const n of cellid.edgeNeighbors(id)) {
      const token = cellid.toToken(n);
      if (!visited.has(token)) {
        visited.add(token);
        queue.push(n);
      }
    }
  }
}

const LEVELS = Array.from({ length: MAX_LEVEL + 1 }, (_, lv) => ({
  value: lv,
  label: `level ${lv}`,
}));

export const s2Adapter: GridAdapter = {
  system: "s2",
  displayName: "S2",
  levels: LEVELS,
  defaultLevel: 13,

  pointToCell(lon, lat, level) {
    const leaf = cellid.fromLatLng(LatLng.fromDegrees(lat, lon));
    return buildCell(cellid.parent(leaf, Number(level)));
  },

  cellToGeometry(id) {
    return buildCell(tokenToId(id));
  },

  getParent(id) {
    const cid = tokenToId(id);
    const level = cellid.level(cid);
    return level > 0 ? cellid.toToken(cellid.parent(cid, level - 1)) : null;
  },

  getChildren(id) {
    const cid = tokenToId(id);
    if (cellid.level(cid) >= MAX_LEVEL) return [];
    return cellid.children(cid).map((c: bigint) => cellid.toToken(c));
  },

  getNeighbors(id) {
    return cellid
      .edgeNeighbors(tokenToId(id))
      .map((c: bigint) => cellid.toToken(c));
  },

  cellsForBounds(bounds, level) {
    const [minLon, minLat, maxLon, maxLat] = bounds;
    const collected = new Map<string, GridCell>();
    const boxes: [number, number, number, number][] =
      minLon <= maxLon
        ? [[minLon, minLat, maxLon, maxLat]]
        : [
            // 国際日付変更線をまたぐ範囲は分割する
            [minLon, minLat, 180, maxLat],
            [-180, minLat, maxLon, maxLat],
          ];
    for (const box of boxes) {
      cellsForBox(box, Number(level), collected, 100000);
    }
    return [...collected.values()];
  },
};
